src/hooks/useTransactionList.ts

```diff
@@ -0,0 +1,152 @@
+  // It resets the stream to the beginning each time it runs,
+  // so we don't want it to self-trigger.
+  React.useEffect(() => {
+    // Reset our mutable state to an empty list:
+    let atEnd = false
+    const changedTxs = new Map<string, EdgeTransaction>()
+    const txs: EdgeTransaction[] = []
+
+    // Sends the mutable state to React,
+    // merging the two transaction lists together:
+    function requestRender() {
+      setOutput({
+        atEnd,
+        transactions: txs.map(tx => changedTxs.get(tx.txid) ?? tx)
+      })
+    }
+    requestRender()
+
+    // If the transaction list changes, we should restart the stream,
+    // but don't delete the transactions we have so far.
+    // Instead, just start re-writing the mutable array until
+    // we get to the end.
+    const cleanupNew = wallet.on('newTransactions', txs => {
+      let relevant = false
+      for (const tx of txs) {
+        if (tx.currencyCode === currencyCode) {
+          relevant = true
+        }
+      }
+      if (relevant) restartStream()
+    })
+
+    // Don't restart the stream if a transaction changes,
+    // but just overlay the new transactions over the old ones:
+    const cleanupChanged = wallet.on('transactionsChanged', txs => {
+      let relevant = false
+      for (const tx of txs) {
+        if (tx.currencyCode === currencyCode) {
+          relevant = true
+          changedTxs.set(tx.txid, tx)
+        }
+      }
+      if (relevant) requestRender()
+    })
+
+    // Constructs a new stream, then updates `cleanupStream` and `requestMore`:
+    let cleanupStream = () => {}
+    function restartStream() {
+      // If this gets set to `true`, the current stream is closed,
+      // so we shouldn't do any more state updates,
+      // even if a lingering promise resolves:
+      let closed = false
+
+      // Keep track of how many transactions we have read,
+      // so we can insert them at the right place:
+      let offset = 0
+
+      // We have a pending promise if this is true, so don't call `next`:
+      let busy = false
+
+      const stream = wallet.streamTransactions({
+        batchSize: 30,
+        firstBatchSize: 10,
+        searchString,
+        tokenId
+      })
+
+      cleanupStream()
+      cleanupStream = () => {
+        closed = true
+        if (stream.return != null) stream.return()
+      }
+
+      requestMore.current = () => {
+        if (busy) return
+        busy = true
+        stream.next().then(
+          result => {
+            busy = false
+            if (closed) return
+            if (result.done) {
+              atEnd = true
+              txs.splice(offset, Infinity) // Trim stragglers
+              requestRender()
+            } else {
+              const { value } = result
+              txs.splice(offset, value.length, ...value)
+              offset += value.length
+              requestRender()
+              if (offset < txs.length) requestMore.current()
```

Comment:
Why are we calling the `requestMore` recursively if the offset doesn't match the length of the result set? Shouldn't this always match; when would it not match?

Reply:
This is for the case were a new transaction arrives from the engine, and we start re-writing the existing transaction list from the beginning. We need to keep going until we reach the end of the existing list.